[Executorch][Export][3/N] Modularize export recipes

abhinaykukkadapu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #12938
* #12937
* #12936

Addresses (7) in the rfc: https://github.com/pytorch/executorch/issues/12660

Changes:
1. Add data class called `LoweringRecipe`
2. Modify current xnnpack recipes to use lowering recipes

Fixes: #12933

Differential Revision: [D79120575](https://our.internmc.facebook.com/intern/diff/D79120575/)

abhinaykukkadapu committed:
[Executorch][Export][3/N] Modularize export recipes

Addresses (7) in the rfc: https://github.com/pytorch/executorch/issues/12660

Changes:
1. Add data class called `LoweringRecipe`
2. Modify current xnnpack recipes to use lowering recipes

Fixes: #12933

Differential Revision: [D79120575](https://our.internmc.facebook.com/intern/diff/D79120575/)

[ghstack-poisoned]

abhinaykukkadapu committed:
Update on "[Executorch][Export][3/N] Modularize export recipes"


Addresses (7) in the rfc: https://github.com/pytorch/executorch/issues/12660

Changes:
1. Add data class called `LoweringRecipe`
2. Modify current xnnpack recipes to use lowering recipes

Fixes: #12933

Differential Revision: [D79120575](https://our.internmc.facebook.com/intern/diff/D79120575/)

[ghstack-poisoned]

digantdesai reviewed (approved):
Stamping to unblock. Left some comments.

abhinaykukkadapu committed:
Update on "[Executorch][Export][3/N] Modularize export recipes"


Addresses (7) in the rfc: https://github.com/pytorch/executorch/issues/12660

Changes:
1. Add data class called `LoweringRecipe`
2. Modify current xnnpack recipes to use lowering recipes

Fixes: #12933

Differential Revision: [D79120575](https://our.internmc.facebook.com/intern/diff/D79120575/)

[ghstack-poisoned]

abhinaykukkadapu commented on export/recipe.py:
```diff
@@ -87,6 +87,26 @@ def get_quantizers(self) -> Optional[List[Quantizer]]:
 
 
+@dataclass
+class LoweringRecipe:
```
@digantdesai I want this to work with either to_edge -> to_backend or to_edge_transform_lower, and broadly we are lowering here, do you think `EdgeLoweringRecipe` make sense?

abhinaykukkadapu commented on export/export.py:
```diff
@@ -189,18 +193,42 @@ def _build_default_stages(self) -> Dict[StageType, Stage]:
                 stage = EdgeTransformAndLowerStage(
-                    partitioners=self._export_recipe.partitioners,
-                    transform_passes=self._export_recipe.edge_transform_passes,
-                    compile_config=self._export_recipe.edge_compile_config,
+                    partitioners=(
+                        self._lowering_recipe.partitioners
+                        if self._lowering_recipe
+                        else None
+                    ),
+                    transform_passes=(
+                        self._lowering_recipe.edge_transform_passes
+                        if self._lowering_recipe
+                        else None
+                    ),
+                    compile_config=(
+                        self._lowering_recipe.edge_compile_config
+                        if self._lowering_recipe
+                        else None
+                    ),
```
@digantdesai  nice idea, will update.

abhinaykukkadapu committed:
Update on "[Executorch][Export][3/N] Modularize export recipes"


Addresses (7) in the rfc: https://github.com/pytorch/executorch/issues/12660

Changes:
1. Add data class called `LoweringRecipe`
2. Modify current xnnpack recipes to use lowering recipes

Fixes: #12933

Differential Revision: [D79120575](https://our.internmc.facebook.com/intern/diff/D79120575/)

[ghstack-poisoned]

abhinaykukkadapu committed:
Update on "[Executorch][Export][3/N] Modularize export recipes"


Addresses (7) in the rfc: https://github.com/pytorch/executorch/issues/12660

Changes:
1. Add data class called `LoweringRecipe`
2. Modify current xnnpack recipes to use lowering recipes

Fixes: #12933

Differential Revision: [D79120575](https://our.internmc.facebook.com/intern/diff/D79120575/)

[ghstack-poisoned]
